src/sentry/hybridcloud/tasks/deliver_webhooks.py

```diff
@@ -159,6 +165,119 @@ def drain_mailbox(payload_id: int) -> None:
+)
+def drain_mailbox_parallel(payload_id: int) -> None:
+    """
+    Deliver messages from a mailbox in small parallel batches.
+
+    Parallel delivery sacrifices strict ordering for increased throughput.
+    Because of the sequential delivery in a mailbox we can't get higher throughput
+    by scheduling batches in parallel.
+
+    Messages will be delivered in small batches until one fails, the batch
+    delay timeout is reached, or a message with a schedule_for greater than
+    the current time is encountered. A message with a higher schedule_for value
+    indicates that we have hit the start of another batch that has been scheduled.
+    """
+    try:
+        payload = WebhookPayload.objects.get(id=payload_id)
+    except WebhookPayload.DoesNotExist:
+        # We could have hit a race condition. Since we've lost already return
+        # and let the other process continue, or a future process.
+        metrics.incr("hybridcloud.deliver_webhooks.delivery", tags={"outcome": "race"})
+        logger.info(
+            "deliver_webhook_parallel.potential_race",
+            extra={
+                "id": payload_id,
+            },
+        )
+        return
+
+    worker_threads = options.get("hybridcloud.webhookpayload.worker_threads")
+    deadline = timezone.now() + BATCH_SCHEDULE_OFFSET
+    request_failed = False
+
+    while True:
+        # We have run until the end of our batch schedule delay. Break the loop so this worker can take another
+        # task.
+        if timezone.now() >= deadline:
+            logger.info(
+                "deliver_webhook_parallel.delivery_deadline",
+                extra={
+                    "mailbox_name": payload.mailbox_name,
+                },
+            )
+            metrics.incr(
+                "hybridcloud.deliver_webhooks.delivery", tags={"outcome": "delivery_deadline"}
+            )
+            break
+
+        # Fetch records from the batch in batch_size blocks. This avoids reading
+        # redundant data should we hit an error and should help keep query duration low.
+        query = WebhookPayload.objects.filter(
+            id__gte=payload.id, mailbox_name=payload.mailbox_name
+        ).order_by("id")
+
+        # No more messages to deliver
+        if query.count() < 1:
+            break
+
+        # Use a threadpool to send requests concurrently
+        with ThreadPoolExecutor(max_workers=worker_threads) as threadpool:
+            futures = {
+                threadpool.submit(deliver_message_parallel, record)
+                for record in query[:worker_threads]
+            }
+            for future in as_completed(futures):
+                payload_record, err = future.result()
+
+                if err:
+                    # Was this the final attempt? Failing on a final attempt shouldn't stop
+                    # deliveries as we won't retry
+                    if payload_record.attempts >= MAX_ATTEMPTS:
+                        payload_record.delete()
```

Comment:
This is a behavior difference from the single-threaded delivery. I was trying to avoid iterating records multiple times and having more complex logic. The downside is sending an 11th request.